### PR TITLE
TiKV update reserve-space config-param

### PR DESCRIPTION
**Pull request by tier-cap**
### What is changed, added or deleted?

1. Update the `storage.reserve-space` description.
2. We have introduced a two-level defense mechanism to protect TiKV storage from being directly overwhelmed by business traffic. This mechanism reuses the `reserve-space` configuration parameter, but the semantics and behavior of this parameter have changed and need to be updated. 

### Which TiDB version(s) do your changes apply to?
- [x] master (the latest development version)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: 
    TiKV: https://github.com/tikv/tikv/pull/10562
    TiDB: https://github.com/pingcap/tidb/pull/26302
    Client-go: https://github.com/tikv/client-go/pull/256
    KVProto: https://github.com/pingcap/kvproto/pull/788
    TiKV: https://github.com/tikv/tikv/pull/10671

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tikv-configuration-file.md [hunk 1]
```diff
@@ -281,7 +281,7 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
++ TiKV 启动时会预留一块磁盘空间用作存储防护, 预留大小 Reserver_Size 取值方法为 Max(Min(disk_total_capacity, `raftstore.capacity`), `reserve-space`)。该预留空间用于防护磁盘被业务流量直接打爆。当磁盘可用空间不足 Reserver_Size 时，业务正常写操作因此失败时会返回 Disk_Full 错误码，此时需要运维干预进行扩容，亦或者执行 Drop Table 释放空间。该预留空间中20%会用作磁盘占位符，名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防犯极端场景下打穿防护空间而导致 Crash 的情况。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。用户配置本参数为 `0MB` 时，防护功能不起作用。
```
Comment (5kbpers) [inline tikv-configuration-file.md:284]:
```suggestion
+ TiKV 启动时会预留一块磁盘空间用作存储防护, 预留大小 Reserver_Size 取值方法为 Max(Min(disk_total_capacity, `raftstore.capacity`), `reserve-space`)。该预留空间用于防护磁盘被业务流量直接打爆。当磁盘可用空间不足 Reserver_Size 时，业务正常写操作因此失败时会返回 Disk_Full 错误码，此时需要运维干预进行扩容，亦或者执行 Drop Table 释放空间。该预留空间中20%会用作磁盘占位符，名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防范极端场景下打穿防护空间而导致 Crash 的情况。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。用户配置本参数为 `0MB` 时，防护功能不起作用。
```

**Review by 5kbpers — APPROVED** [review]

File: tikv-configuration-file.md [hunk 2]
```diff
@@ -281,7 +281,7 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
++ TiKV 启动时会预留一块磁盘空间用作存储防护, 预留大小 Reserver_Size 取值方法为 Max(Min(disk_total_capacity, `raftstore.capacity`), `reserve-space`)。该预留空间用于防护磁盘被业务流量直接打爆。当磁盘可用空间不足 Reserver_Size 时，业务正常写操作因此失败时会返回 Disk_Full 错误码，此时需要运维干预进行扩容，亦或者执行 Drop Table 释放空间。该预留空间中20%会用作磁盘占位符，名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防范极端场景下打穿防护空间而导致 Crash 的情况。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。用户配置本参数为 `0MB` 时，防护功能不起作用。
```
Comment (hicqu) [inline tikv-configuration-file.md:284]: s/Reverser/Reverve

**Comment by hicqu** [comment]
Rest LGTM.

File: tikv-configuration-file.md (hunk 2 above)
Comment (tier-cap) [reply tikv-configuration-file.md]: Done

**Review by hicqu — APPROVED** [review]

File: tikv-configuration-file.md [hunk 3]
```diff
@@ -281,7 +281,7 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
++ TiKV 启动时会预留一块磁盘空间用作存储防护，预留大小 Reserve_Size 取值方法为 Max(Min(disk_total_capacity, `raftstore.capacity`), `reserve-space`)。该预留空间用于防护磁盘被业务流量直接打爆。当磁盘可用空间不足 Reserve_Size 时，业务正常写操作因此失败时会返回 Disk_Full 错误码，此时需要运维干预进行扩容，亦或者执行 Drop Table 释放空间。该预留空间中20%会用作磁盘占位符，名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下打穿防护空间而导致 Crash 的情况。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。用户配置本参数为 `0MB` 时，防护功能不起作用。
```
Comment (qiancai) [inline tikv-configuration-file.md:284]:
```suggestion
+ TiKV 启动时会预留一块磁盘空间，用于存储防护。此配置项将和`raftstore.capacity` 以及磁盘总空间（disk_total_capacity）共同控制 TiKV 的预留空间 Reserve_Size。Reserve_Size 的取值方法为 Max(Min(disk_total_capacity, `raftstore.capacity`), `reserve-space`)。TiKV 预留空间的 20% 会作为磁盘占位符，名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下磁盘空间被写满导致 TiKV 重启。

    + 当磁盘可用空间低于 Reserve_Size 导致业务正常写操作失败时，TiKV 会返回 Disk_Full 错误码，提示失败原因。此时，可以通过扩容，或者执行 Drop Table 释放磁盘空间。
    + 当 TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除 `space_placeholder_file`，并且将 `reserve-space` 设置为 `0MB` （代表关闭防护功能）。
```

Comment (qiancai) [reply tikv-configuration-file.md:284]: 在设置本参数为 0 MB 之后，当 TiKV 磁盘恢复时，是否建议用户将此参数设置回 5 MB 呢？

Comment (qiancai) [reply tikv-configuration-file.md:284]: 找了下https://docs.pingcap.com/zh/tidb/stable/tikv-configuration-file，没有找到 raftstore.capacity 这个配置项，可以帮忙看下吗

**Review by qiancai — COMMENTED** [review]
@tier-cap PTAL

File: tikv-configuration-file.md [hunk 4]
```diff
@@ -281,7 +281,7 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
++ TiKV 启动时会预留一块磁盘空间用作存储防护，预留空间大小为部署 TiKV 的存储的容量的百分之五与本配置参数之间的最大值。该预留空间用于防护磁盘被业务流量直接打爆。当磁盘可用空间不足预留空间时，业务正常写操作失败会返回 Disk_Full 错误码，此时需要运维干预进行扩容，亦或者执行 Drop Table 释放空间。该预留空间中80%空间用作软防御，20%会用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下打穿软防御空间而导致 Crash 的情况。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。待 TiKV 实例的存储空间余额充沛后建议将该参数重新设置，以启用磁盘防护功能。用户配置本参数为 `0MB` 时，防护功能不起作用。
```
Comment (qiancai) [inline tikv-configuration-file.md:284]:
```suggestion
+ TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。

- 当磁盘可用空间不足预留空间时，业务正常写操作失败会返回 Disk_Full 错误码。此时需要运维干预进行扩容，或者执行 Drop Table 释放空间。
- 当 TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除 `space_placeholder_file` 文件，并且将 `reserve-space` 设置为 `0MB`（代表关闭软防御防护功能）。待 TiKV 实例的存储空间充足后，建议重新设置该参数，以启用磁盘防护功能。
```

Comment (tier-cap) [reply tikv-configuration-file.md:284]: （代表关闭软防御防护功能） 需要改成 （代表关闭磁盘防护功能）
因为：80%为软防御，20%为硬防御

File: tikv-configuration-file.md [hunk 5]
```diff
@@ -281,7 +281,10 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
++ TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。
```
Comment (qiancai) [inline tikv-configuration-file.md:284]:
```suggestion
TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。
```

File: tikv-configuration-file.md [hunk 6]
```diff
@@ -281,7 +281,10 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
+TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。
+
+- 当磁盘可用空间不足预留空间时，业务正常写操作失败会返回 Disk_Full 错误码。此时需要运维干预进行扩容，或者执行 Drop Table 释放空间。
+- 当 TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除 `space_placeholder_file` 文件，并且将 `reserve-space` 设置为 `0MB`（代表关闭磁盘防护功能）。待 TiKV 实例的存储空间充足后，建议重新设置该参数，以启用磁盘防护功能。
```
Comment (qiancai) [inline tikv-configuration-file.md:287]:
```suggestion
- TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。

    + 当磁盘可用空间不足预留空间时，业务正常写操作失败会返回 Disk_Full 错误码。此时需要运维干预进行扩容，或者执行 Drop Table 释放空间。
    + 当 TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除 `space_placeholder_file` 文件，并且将 `reserve-space` 设置为 `0MB`（代表关闭磁盘防护功能）。待 TiKV 实例的存储空间充足后，建议重新设置该参数，以启用磁盘防护功能。
    
```

File: tikv-configuration-file.md [hunk 7]
```diff
@@ -281,7 +281,10 @@ TiKV 配置文件比命令行参数支持更多的选项。你可以在 [etc/con
 
 ### `reserve-space`
 
-+ TiKV 启动时预占额外空间的临时文件大小。临时文件名为 `space_placeholder_file`，位于 `storage.data-dir` 目录下。TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除该文件，并且将 `reserve-space` 设置为 `0MB`。
++ TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。
+
+- 当磁盘可用空间不足预留空间时，业务正常写操作失败会返回 Disk_Full 错误码。此时需要运维干预进行扩容，或者执行 Drop Table 释放空间。
+- 当 TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除 `space_placeholder_file` 文件，并且将 `reserve-space` 设置为 `0MB`（代表关闭磁盘防护功能）。待 TiKV 实例的存储空间充足后，建议重新设置该参数，以启用磁盘防护功能。
```
Comment (qiancai) [inline tikv-configuration-file.md:287]:
```suggestion
+ TiKV 启动时会预留一块磁盘空间用于防护磁盘被写满。预留空间大小为 TiKV 存储容量的百分之五与本配置参数之间的最大值。预留空间中 80% 的空间用作软防御，20% 的空间用作磁盘占位符。占位符名称为 `space_placeholder_file`，位于 `storage.data-dir` 目录下，用于防止极端场景下软防御空间也被写满而导致 TiKV Crash。

    - 当磁盘可用空间不足预留空间时，业务正常写操作失败会返回 Disk_Full 错误码。此时需要运维干预进行扩容，或者执行 Drop Table 释放空间。
    - 当 TiKV 磁盘空间耗尽无法正常启动需要紧急干预时，可以删除 `space_placeholder_file` 文件，并且将 `reserve-space` 设置为 `0MB`（代表关闭磁盘防护功能）。待 TiKV 实例的存储空间充足后，建议重新设置该参数，以启用磁盘防护功能。
    
```

**Review by qiancai — APPROVED** [review]
LGTM

**Pull request merged**